src/Interpreters/ProcessList.cpp

```diff
@@ -428,25 +436,72 @@ QueryStatus * ProcessList::tryGetProcessListElement(const String & current_query
 }
 
 
-CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id, const String & current_user, bool kill)
+CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id, const String & current_user)
 {
-    auto lock = safeLock();
+    auto [lock, blocker] = safeLock();
 
     QueryStatus * elem = tryGetProcessListElement(current_query_id, current_user);
+    // To forbid element to be removed from process list
+    elem->forbidRemoval();
+    lock.unlock();
 
     if (!elem)
         return CancellationCode::NotFound;
 
-    return elem->cancelQuery(kill);
+    auto code =  elem->cancelQuery();
+    lock.lock();
+    // If query already finished we need to remove it from process list
+    if (elem->isReadyForRemoval())
```

Comment:
Is there guarantee that `sendCancelToQuery` will be called for the query with `KillQueryFlags::READY_FOR_REMOVAL`?

Actually AFAICS `cancelQuery` can be already called w/o the lock, since it already has atomic flag, am I missing something?

Reply:
>Actually AFAICS cancelQuery can be already called w/o the lock, since it already has atomic flag, am I missing something?

Will answer by myself, it is not that simple, since you need to ensure that the point is still valid.

I've looked into this and came with a simple patch - #41467, @novikd what do you think about it?